src/api/endpoint/api/star.js

```diff
@@ -0,0 +1,49 @@
+// @flow
+
+import type {$Response} from 'express';
+import type {$RequestExtend, $NextFunctionVer, IStorageHandler} from '../../../../types';
+
+export default function(storage: IStorageHandler) {
+  return (req: $RequestExtend, res: $Response, next: $NextFunctionVer): void => {
+    const name = req.params.package;
+    // Check is star or unstar
+    const isStar = Object.keys(req.body.users).indexOf(req.remote_user.name) >= 0;
+    const afterChangePackage = function(err) {
+      if (err) {
+        return next(err);
+      }
+      res.status(200);
+      next({
+        success: true,
+      });
+    };
+
+    storage.getPackage({
+      name,
+      req,
+      callback: function(err, info) {
+        if (err) {
+          return next(err);
+        }
+        let newUsers = {
+          ...info.users,
+        };
+        if (isStar) {
+          newUsers = {
+            ...newUsers,
+            [req.remote_user.name]: true,
+          };
+        } else if (newUsers[req.remote_user.name]) {
+          delete newUsers[req.remote_user.name];
```

Comment:
```
delete newUsers[req.remote_user.name];
```
I'd rather propose this model, it avoids write into the file system whether is not necesary, avoid use delete in favor if reducer. No mutations. What do you think? Perhaps you can improve it.

```
// @flow

import type {$Response} from 'express';
import _ from 'lodash';
import type {$RequestExtend, $NextFunctionVer, IStorageHandler} from '../../../../types';

const validateInputs = (newUsers, localUsers, username, isStar) => {
  const isExistlocalUsers = _.isNil(localUsers[username]) === false;
  
  if (isStar && isExistlocalUsers && localUsers[username]) {
    return true;
  } else if (!isStar && isExistlocalUsers) {
    return false;
  } else if (!isStar && !isExistlocalUsers) {
    return true;
  } else {
    return false;
  }
}

export default function(storage: IStorageHandler) {
  return (req: $RequestExtend, res: $Response, next: $NextFunctionVer): void => {
    const name = req.params.package;
    const afterChangePackage = function(err) {
      if (err) {
        return next(err);
      }
      res.status(200);
      next({
        success: true,
      });
    };

    storage.getPackage({
      name,
      req,
      callback: function(err, info) {
        if (err) {
          return next(err);
        }

        const newStarUser = req.body.users;
        const remoteUsername = req.remote_user.name;
        const isStar = Object.keys(newStarUser).includes(remoteUsername);

        if (_.isNil(info.users) === false && validateInputs(newStarUser, info.users, remoteUsername, isStar)) {
            return afterChangePackage();
        }

        const users = isStar ? {
          ...info.users,
          [remoteUsername]: true,
        } : _.reduce(info.users, (users, value, key) => {
            if (key !== remoteUsername) {
              users[key] = value;
            }

            return users;
        }, {});

        storage.changePackage(name, {...info, users}, req.body._rev, function(err) {
          return afterChangePackage(err);
        });
      },
    });
  };
}

````